docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,161 @@
+---
+title: auto-return-agent-to-infra-env
+authors:
+  - "@carbonin"
+creation-date: 2022-05-17
+last-updated: 2022-05-26
+---
+
+# Automatically Return Agents to InfraEnv
+
+## Summary
+
+Without a BareMetalHost there is currently no way to automate the process
+of unbinding a host from a cluster. When a user wants to reuse a host for a new
+cluster they have to use the assisted-service unbind API, then manually boot
+the host back into the discovery image.
+
+To support automating this procedure, the assisted-service will configure nodes
+in installed clusters to run an agent. When appropriate this agent will
+reconfigure the host to boot into a discovery image, allowing the host to
+be made available for installation without the user manually interacting with
+a management system to reboot the host into the discovery image.
+
+## Motivation
+
+A user creating and managing clusters using the assisted-service may not have
+access to the systems used to configure and boot hosts. This is especially
+true in bare-metal use cases where a BareMetalHost is not present. The
+assisted-service can make this use case easier by handling the reprovisioning
+process.
+
+### Goals
+
+* Reconfigure a host to boot the discovery image as a part of unbinding from a
+  cluster without a BareMetalHost
+* Allow for a node to be gracefully removed from an installed cluster
+
+### Non-Goals
+
+* Automate graceful removal of a node from an existing cluster
+* Booting a node that is unresponsive or not currently part of an installed
+  cluster
+* Registering an agent to an InfraEnv other than the one it is currently
+  associated with
+
+## Proposal
+
+When a host should be returned to the InfraEnv use the kubeconfig created when
+the cluster was installed to run an agent on the host. This agent will handle
+downloading the required artifacts and rebooting the host into the discovery
+process.
+
+This process will only be used in situations where the assisted-service is sure
+to have connectivity to the installed cluster's API endpoint. In cases such as
+the SaaS this feature will not be used.
```

Comment:
I think I understand why it won't be used in SaaS, but still might be worth mentioning it here.

Reply:
I did, right?

>This process will only be used in situations where the assisted-service is sure to have connectivity to the installed cluster's API endpoint

Reply:
I did, right?

>This process will only be used in situations where the assisted-service is sure to have connectivity to the installed cluster's API endpoint